tests/Unit/CrudPanel/CrudPanelButtonsTest.php

```diff
@@ -0,0 +1,218 @@
+
+        $this->assertEquals(count($this->defaultButtonNames) + 1, count($this->crudPanel->buttons));
+    }
+
+    public function testAddButtonBeginning()
+    {
+        $expectedButton = $this->topViewButton;
+
+        $this->crudPanel->addButton($expectedButton->stack, $expectedButton->name, $expectedButton->type, $expectedButton->content, 'beginning');
+
+        $this->assertEquals($expectedButton, $this->crudPanel->buttons->first());
+    }
+
+    public function testAddButtonEnd()
+    {
+        $expectedButton = $this->lineViewButton;
+
+        $this->crudPanel->addButton($expectedButton->stack, $expectedButton->name, $expectedButton->type, $expectedButton->content, 'end');
+
+        $this->assertEquals($expectedButton, $this->crudPanel->buttons->last());
+    }
+
+    public function testAddButtonUnknownPosition()
+    {
+        $this->markTestIncomplete('Not correctly implemented');
+
+        $this->setExpectedException(\Exception::class);
+
+        $expectedButton = $this->lineViewButton;
+
+        // TODO: this should throw an error.
+        $this->crudPanel->addButton($expectedButton->stack, $expectedButton->name, $expectedButton->type, $expectedButton->content, 'unknownPosition');
+    }
+
+    public function testAddButtonFromModelFunction()
+    {
+        $expectedButton = $this->topModelFunctionButton;
+
+        $this->crudPanel->addButton($expectedButton->stack, $expectedButton->name, $expectedButton->type, $expectedButton->content);
+
+        $this->assertEquals($expectedButton, $this->crudPanel->buttons->last());
+    }
+
+    public function testAddButtonFromView()
+    {
+        $expectedButton = $this->topViewButton;
+        $viewName = 'someViewName';
+
+        $this->crudPanel->addButtonFromView($expectedButton->stack, $expectedButton->name, $viewName, $expectedButton->content);
+
+        $backpackButtonViewPackage = 'vendor.backpack.crud.buttons.';
+        $actualButton = $this->crudPanel->buttons->last();
+
+        $this->assertEquals($expectedButton->stack, $actualButton->stack);
+        $this->assertEquals($expectedButton->name, $actualButton->name);
+        $this->assertEquals($backpackButtonViewPackage.$viewName, $actualButton->content);
+        $this->assertEquals($expectedButton->stack, $actualButton->stack);
+    }
+
+    public function testRemoveButton()
+    {
+        $buttonName = $this->defaultButtonNames[0];
+
+        $this->crudPanel->removeButton($buttonName);
+
+        $this->assertEquals(count($this->defaultButtonNames) - 1, count($this->crudPanel->buttons));
+        $this->assertNull($this->getButtonByName($buttonName));
+    }
+
+    public function testRemoveUnknownButton()
+    {
+        $buttonName = 'someButtonName';
+
+        $this->crudPanel->removeButton($buttonName);
+
+        $this->assertEquals(count($this->defaultButtonNames), count($this->crudPanel->buttons));
```

Comment:
Better to check that it's not there at all - what if `removeButton` randomly alters some other button's name or such?

Reply:
True. Will add this.